language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/impl/DiagBasedCodeAction.java

```diff
@@ -0,0 +1,199 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.codeaction.impl;
+
+import io.ballerina.compiler.api.types.BallerinaTypeDescriptor;
+import org.ballerinalang.langserver.common.ImportsAcceptor;
+import org.ballerinalang.langserver.common.utils.FunctionGenerator;
+import org.ballerinalang.langserver.commons.LSContext;
+import org.ballerinalang.langserver.commons.codeaction.LSCodeActionProviderException;
+import org.ballerinalang.langserver.compiler.DocumentServiceKeys;
+import org.ballerinalang.model.tree.TopLevelNode;
+import org.ballerinalang.model.tree.expressions.ExpressionNode;
+import org.ballerinalang.model.tree.expressions.RecordLiteralNode;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.TextEdit;
+import org.wso2.ballerinalang.compiler.semantics.analyzer.Types;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BRecordType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTupleType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.compiler.tree.BLangPackage;
+import org.wso2.ballerinalang.compiler.tree.BLangTypeDefinition;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangBinaryExpr;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangFieldBasedAccess;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangListConstructorExpr;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangQueryExpr;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral;
+import org.wso2.ballerinalang.compiler.tree.types.BLangRecordTypeNode;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+/**
+ * Interface for diagnostics based code actions.
+ *
+ * @since 2.0.0
+ */
+public interface DiagBasedCodeAction {
+
+    List<CodeAction> get(Diagnostic diagnostic, List<Diagnostic> allDiagnostics, LSContext context)
+            throws LSCodeActionProviderException;
+
+
+    static List<String> getPossibleTypes(LSContext context,
+                                         BallerinaTypeDescriptor typeDescriptor,
+                                         List<TextEdit> edits,
+                                         CompilerContext compilerContext) {
+        ImportsAcceptor importsAcceptor = new ImportsAcceptor(context);
+
+        List<String> types = new ArrayList<>();
+
+        String variableType = FunctionGenerator.generateTypeDefinition(importsAcceptor, typeDescriptor, context);
+        if (typeDescriptor instanceof BLangInvocation) {
```

Comment:
No, this has to revamped when we get `type-info` for literals. Kept it intact since for the reference of code-migration and it doesn't affect the other flow. Should we remove these checks?